src/database/database-files.cpp

```diff
@@ -177,3 +178,113 @@ void PlayerDatabaseFiles::listPlayers(std::vector<std::string> &res)
 		res.emplace_back(player.getName());
 	}
 }
+
+AuthDatabaseFiles::AuthDatabaseFiles(const std::string &savedir) : m_savedir(savedir)
+{
+	readAuthFile();
+}
+
+bool AuthDatabaseFiles::getAuth(const std::string &name, AuthEntry &res)
+{
+	const auto res_i = m_auth_list.find(name);
+	if (res_i == m_auth_list.end()) {
+		return false;
+	}
+	res = res_i->second;
+	return true;
+}
+
+bool AuthDatabaseFiles::saveAuth(const AuthEntry &authEntry)
+{
+	m_auth_list[authEntry.name] = authEntry;
+
+	// save entire file
+	return writeAuthFile();
+}
+
+bool AuthDatabaseFiles::createAuth(AuthEntry &authEntry)
+{
+	m_auth_list[authEntry.name] = authEntry;
+
+	// save entire file
+	return writeAuthFile();
+}
+
+bool AuthDatabaseFiles::deleteAuth(const std::string &name)
+{
+	auto count = m_auth_list.erase(name);
+	if (!count) {
+		// did not delete anything -> hadn't existed
+		return false;
+	}
+	return writeAuthFile();
+}
+
+void AuthDatabaseFiles::listNames(std::vector<std::string> &res)
+{
+	res.clear();
+	res.reserve(m_auth_list.size());
+	for (auto const &res_pair : m_auth_list) {
+		res.push_back(res_pair.first);
+	}
+	return;
+}
+
+void AuthDatabaseFiles::reload()
+{
+	readAuthFile();
+	return;
+}
+
+bool AuthDatabaseFiles::readAuthFile()
+{
+	std::string path = m_savedir + DIR_DELIM + "auth.txt";
+	std::ifstream file(path, std::ios::binary);
+	if (!file.good()) {
+		return false;
+	}
+	m_auth_list.clear();
+	while (file.good()) {
+		std::string line;
+		std::getline(file, line);
+		std::vector<std::string> parts = str_split(line, ':');
+		if (parts.size() < 4) // also: empty line at end
+			continue;
+		std::string &name = parts[0];
+		std::string &password = parts[1];
+		std::vector<std::string> privileges = str_split(parts[2], ',');
+		time_t last_login = atol(parts[3].c_str());
+
+		AuthEntry authEntry = {
+				1,
+				name,
+				password,
+				privileges,
+				last_login,
+		};
+		m_auth_list[name] = authEntry;
+	}
+	return true;
+}
+
+bool AuthDatabaseFiles::writeAuthFile()
+{
+	std::string path = m_savedir + DIR_DELIM + "auth.txt";
+	std::ostringstream output(std::ios_base::binary);
+	for (const auto &auth_i : m_auth_list) {
+		const AuthEntry &authEntry = auth_i.second;
+		output << authEntry.name << ":" << authEntry.password << ":";
+		output << *(authEntry.privileges.begin());
+		for (auto p = std::next(authEntry.privileges.begin());
```

Comment:
for (const auto &p: authEntry.privileges)

Reply:
He implements his own "vector concatenation" function here, which doesn't lead to a superfluous beginning/tailing comma. Because that one would result in an empty string in `str_split`, which is not desired as a priv value.